File: mcp/bifract_mcp/bql.py
"""Helpers for assembling BQL from tool arguments."""

BREAKOUT_CHARS = '"\'\\|'


class BQLValueError(ValueError):
    """A tool argument cannot be placed into a query safely."""


def time_window(start: str, end: str) -> dict:
    """Build the start/end fragment of a query request body, omitting blanks."""
    window = {}
    if start.strip():
        window["start"] = start.strip()
    if end.strip():
        window["end"] = end.strip()
    return window


def literal(value: str, *, field: str) -> str:
    """Quote a value for a BQL comparison.

    BQL has no escape sequence inside a quoted value, so a value carrying a quote,
    backslash, pipe, or control character cannot be represented and is rejected rather
    than silently changing the query's meaning.
    """
    if any(c in value for c in BREAKOUT_CHARS) or any(ord(c) < 0x20 for c in value):
        raise BQLValueError(
            f"{field} contains a character BQL cannot quote "
            f"(one of {BREAKOUT_CHARS} or a control character): {value!r}"
        )
    return f'"{value}"'


def contains(field: str, value: str) -> str:
    """A case-insensitive substring match (=~), which is index-accelerated."""
    return f"{field}=~{literal(value, field=field)}"


def equals(field: str, value: str) -> str:
    return f"{field}={literal(value, field=field)}"

File: pkg/aitools/bql.go
package aitools

import "fmt"

// breakoutChars cannot appear inside a BQL quoted value.
const breakoutChars = `"'\|`

// bqlLiteral quotes a value for a BQL comparison. BQL has no escape sequence
// inside a quoted value, so anything that would break out is rejected rather
// than silently changing the query's meaning.
func bqlLiteral(value, field string) (string, error) {
	for _, r := range value {
		if r < 0x20 {
			return "", fmt.Errorf("%s contains a control character, which BQL cannot quote: %q", field, value)
		}
		for _, bad := range breakoutChars {
			if r == bad {
				return "", fmt.Errorf(
					"%s contains a character BQL cannot quote (one of %s): %q", field, breakoutChars, value)
			}
		}
	}
	return `"` + value + `"`, nil
}

// bqlContains is a case-insensitive substring match, which is index-accelerated.
func bqlContains(field, value string) (string, error) {
	literal, err := bqlLiteral(value, field)
	if err != nil {
		return "", err
	}
	return field + "=~" + literal, nil
}

func bqlEquals(field, value string) (string, error) {
	literal, err := bqlLiteral(value, field)
	if err != nil {
		return "", err
	}
	return field + "=" + literal, nil
}
